### PR TITLE
Allow concurrent pushes within the same repository

### LINKED ISSUE #9132: Can't push multiple images to same repo in parallel
Attempting to push multiple images to the same remote in parallel cases one of the push operations to fail. The operation claims "is already in progress", implying that the image is already being pushed, but this is false. The image is not pushed, and pushing again at a later time will end up actually pushing the image.

```
$ cat > Dockerfile <<'EOF'
FROM fedora:20
ADD file /file
RUN dd if=/dev/zero of=zero bs=1M count=10
EOF

$ echo 1 > file

$ docker build -t phemmer/test:1 .

$ echo 2 > file

$ docker build -t phemmer/test:2 .

$ docker push phemmer/test:1 & docker push phemmer/test:2 ; wait
[1] 31920
The push refers to a repository [phemmer/test] (len: 1)
Sending image list
2014/11/12 20:38:02 Error: push phemmer/test is already in progress
[1]  + exit 1     docker push phemmer/test:1
Pushing repository phemmer/test (1 tags)
511136ea3c5a: Image already pushed, skipping 
c69cab00d6ef: Image already pushed, skipping 
88b42ffd1f7c: Image already pushed, skipping 
64dd2ddab00b: Image successfully pushed 
712a4eb2c2eb: Image successfully pushed 
Pushing tag for rev [712a4eb2c2eb] on {https://cdn-registry-1.docker.io/v1/repositories/phemmer/test/tags/2}
```

In this case, the `phemmer/test:2` image successfully pushed, but `phemmer/test:1` failed.

Docker version 1.3.1

### PR DESCRIPTION
**Pull request by aaronlehmann**
This change brings push behavior in line with pulling. If the requested
push matches one already in progress (same tag, if any), show the
progress of outstanding push operation. Otherwise, a new push is
started. Layer uploads are also deduplicated (but only within the same
repository, to avoid cross-repository push issues).

This change is only being made for v2 pushes.

Fixes #9132

cc: @dmcgowan @tonistiigi @stevvooe

### CODE REVIEW

File: graph/pull_v2.go [hunk 1]
```diff
@@ -75,7 +75,8 @@ func (p *v2Puller) pullV2Repository(tag string) (err error) {
 	broadcaster, found := p.poolAdd("pull", taggedName)
 	broadcaster.Add(p.config.OutStream)
 	if found {
-		// Another pull of the same repository is already taking place; just wait for it to finish
+		// Another push or pull of the same repository is already taking place; just wait
```
Comment (tonistiigi) [inline graph/pull_v2.go]: push or pull? don't they still go to different queues?

Comment (aaronlehmann) [inline graph/pull_v2.go]: The `poolAdd` function checks both queues.

**Comment by aaronlehmann** [comment]
I've changed the logic around pushing at the same time as a pulling:

- If one or more pulls are running in a given repo, trying any push in that repo will wait for all the pulls to finish, then do the push.
- If one or more pushes are running in a given repo, trying any pull in that repo will wait for all the pushes to finish, then do the pull.

I think this is the best approach. We don't want a pull to happen at the same time as a push. I originally tried using the broadcaster pool to implement this, but it didn't work well - push pushes all tags by default, but pull only pulls latest, so the pool keys didn't match and it was possible to push and pull at the same time.

@tonistiigi PTAL

**Comment by tonistiigi** [comment]
Is there a way to combine acquirePull/acquirePush, releasePull/releasePull to a more general function? They seems to have identical functionality.

The code that calls these pulls is not very easy to understand. You have to know everything about the internals to figure out what its doing. I wonder if it would be simpler if we have a simple pool and we first take a lock by the `CanonicalName` and then more specific lock for the `name+action`.

What is the use case of incrementing repo counters when downloading the layers? Shouldn't the tag name lock already take care of the possibility of collision. Is there a way where the wait function would ever be called in these cases(and load).

**Comment by aaronlehmann** [comment]
> Is there a way to combine acquirePull/acquirePush, releasePull/releasePull to a more general function? They seems to have identical functionality.

The problem is that `acquirePull` has to check for outstanding pushes, and `acquirePush` has to check for outstanding pulls, so it doesn't map well to having `push` and `pull` objects with `acquire` and `release` methods. Any ideas?

> The code that calls these pulls is not very easy to understand. You have to know everything about the internals to figure out what its doing. I wonder if it would be simpler if we have a simple pool and we first take a lock by the CanonicalName and then more specific lock for the name+action.

The reason why there are separate maps called `pullsByKey` and `pullCountByRepo` is that for specific keys, we need to be able to find the broadcaster (to attach an identical pull), but we also need to track how many pulls are happening in a repo so we can wait for them all to finish - thus two different value types. I'm not sure I understand the suggestion about using nested locks.

> What is the use case of incrementing repo counters when downloading the layers? Shouldn't the tag name lock already take care of the possibility of collision. Is there a way where the wait function would ever be called in these cases(and load).

It's not necessary when downloading the layers; I just preferred to reuse the existing set of functions instead of adding a new set just for layers.

**Comment by endophage** [comment]
Is the tag the right level of abstraction for determining if 2 pushes are the same? If I try to push the same tag with different content on the second push, I (maybe a minority on this) would want the first push to get cancelled.

**Comment by aaronlehmann** [comment]
Unfortunately, right now there's no way to know in advance whether the content is different. Looking at the tag is the best we can do. In the 1.10 timeframe, what you're suggesting could be implemented.

**Comment by endophage** [comment]
Ahh, I would suggest then that it doesn't make sense to add this feature until 1.10 or later. It just seems a little too best effort right now with a large margin for error.

**Comment by aaronlehmann** [comment]
We could change this patch to wait for the current push to finish and then start a new one. If no data has changed, the second push would complete quickly. This would ensure that the last push wins if the data pointed to by the tag changes.

**Comment by endophage** [comment]
I think that would be a good improvement. It just ensures if the user has made a change and pushes before the first one finishes, there's no confusion about what's actually on the server.

**Comment by aaronlehmann** [comment]
Closing in favor of #16891. @tonistiigi @endophage PTAL

**Pull request closed without merging**